ouroboros-network/src/Ouroboros/Network/Mux/Control.hs

```diff
@@ -0,0 +1,79 @@
+{-# LANGUAGE BangPatterns   #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TypeFamilies   #-}
+
+module Ouroboros.Network.Mux.Control (
+      ControlMsg (..)
+    , encodeCtrlMsg
+    , decodeCtrlMsg
+    ) where
+
+import qualified Codec.CBOR.Decoding as CBOR
+import qualified Codec.CBOR.Encoding as CBOR
+import qualified Codec.CBOR.Read as CBOR
+import qualified Codec.CBOR.Term as CBOR
+import           Codec.CBOR.Write (toLazyByteString)
+import           Codec.Serialise.Class
+import           Control.Exception
+import           Control.Monad
+import qualified Data.Map.Strict as M
+import           Data.Maybe (catMaybes)
+import           GHC.Stack
+
+import           Ouroboros.Network.Mux.Types
+
+data ControlMsg = MsgInitReq (M.Map VersionNumber Version)
+                | MsgInitRsp Version
+                | MsgInitFail String
+                deriving (Eq, Show)
+
+encodeCtrlMsg :: ControlMsg -> CBOR.Encoding
+encodeCtrlMsg (MsgInitReq versions) = CBOR.encodeListLen 2 <> CBOR.encodeWord 0 <> encodeVersionMap versions
+encodeCtrlMsg (MsgInitRsp version)  = CBOR.encodeListLen 2 <> CBOR.encodeWord 1 <> encode version
+encodeCtrlMsg (MsgInitFail msg)     = CBOR.encodeListLen 2 <> CBOR.encodeWord 2 <> encode msg
+
+encodeVersionMap :: M.Map VersionNumber Version -> CBOR.Encoding
+encodeVersionMap versions =
+    CBOR.encodeMapLen (fromIntegral $ M.size versions)
+    <> foldr (\v b -> encode v <> b) mempty (M.elems versions)
+
+decodeCtrlMsg :: HasCallStack => CBOR.Decoder s ControlMsg
+decodeCtrlMsg = do
+    _ <- CBOR.decodeListLen
+    key <- CBOR.decodeWord
+    case key of
+         0 -> MsgInitReq . M.mapMaybeWithKey decodeVersionTerm <$> decodeVersionMap
+         1 -> MsgInitRsp <$> decode
+         2 -> MsgInitFail <$> decode
+         a -> throw $ MuxError MuxControlUnknownMessage
+                      ("unknown control message type " ++ show a)
+                      callStack
+
+  where
+    decodeVersionMap :: CBOR.Decoder s (M.Map VersionNumber CBOR.Term)
+    decodeVersionMap = do
+        n <- CBOR.decodeMapLen
+        let decodeEntry = do
+              !k <- CBOR.decodeWord
+              !v <- CBOR.decodeTerm
```

Comment:
I don't think we need to use `decodeTerm` here.  You could just use `decode` (since `Version` has `Serialise` instance` and thus we could have `decodeVersionMap :: CBOR.Decoder s (M.Map VersionNumber Version)`.

The `decodeVersionMap` and `encodeVersionMap` are not compatible.  The latter is using `encode :: (VersionNumber, Version) -> CBOR.Encoding`, which will be encoded as a list of length `2` with two entries.  This is not reflected in `decodeEntry`.

Reply:
I can't use a plain `decode` since there could be versions which we don't know how to decode and we need to consume all the bytes for that unknown `Term`. This is tested in `prop_unknown_version_req`.

The `decodeVersionMap` and `encodeVersionMap` are compatible, which is demonstrated by `prop_mux_encode_decode`. However the first implementation of `encodeVersionMap` used `map` instead of `foldr` which resulted in exactly the behavior your described, every entry
was encoded as a list with two entries. I lifted the `foldr` trick from `cardano-sl`, and why it works is not at all apparent to me. I couldn't find any documentation or examples of how to encode a Map in the CBOR library itself.

Reply:
Example Messge without a list lenght field:
`0x82 0x00 0xa1 0x00 0x1a 0xca 0xfe 0xbe 0xeb`

Reply:
I got confused by `M.elems`, it actually returns just a list of values that are inside the map, e.g.
```
ghci> foldr (\v b -> encode v <> b) mempty  (M.elems $ M.fromList 
([(1,'a'), (2,'b')] :: [(Int, Char)]))
[TkString "a",TkString "b"]
```
That's why the encoder and decoder match. The only way to encode a map in `cborg` seems to be as a list of tuples.   Both approaches are not really using the `VersionNumber` in the decoder or encoder.

If we'll need to add a new version, then we will need to change the `decodeVersionTerm` and not just `Serialise` instance for `Version`. 

 If we use `[Version]` instead of `Map VersionNumber Version` and validate that there are no duplicates, or even better, do that while we are decoding (so we don't read the whole list `replicate 10000 version0`). We can fail early in this case.

Reply:
The largest possible message we can receive is 64k, i.e limited by a single mux segment, that is early enough. 

When we add a new version ghc will spit out a warning indicating that you need to update `decodeVersionTerm`. I don't see a problem with the current `Map VersionNumber Version` implementation.

Reply:
So why to bother at all with the `Map` over a list?  It seems it only adds complexity.

Reply:
I think that the complexity comes from the fact that we need to be able to skip over unknown versions, not from the fact that we're using a `Map` instead of a list.

Reply:
But with a list and no `VersionNumber` you could equally well skip unknown versions.

Reply:
This is turning into a bike shed.

Reply:
Not really; I just don't see the point of `VersionNumber`.

Cons:
 - you need to come up with a new one when we add a new protocol version
-  we need to adjust `decodeVersionTerm` not just `Serialise` instance of `Version`
Pros:
- it allows us to check we are not sending multiple versions, but it is not preventing from receiving multiple versions (since the encoded format is almost the same as a list of `Version`, the difference is just `TkMapLen` over `TkListLen`). Any anyway mux SDU limit protects us from receiving too much data.

@avieth maybe you'll have a power to break the tie :).

Another issue: I would expect that if we are using `encodeMapLen` then what follows are tuples (`key` and `value`), rather than just values; but somebody with more `cborg` experience should raise his voice here.  Maybe @MarcFontaine has already gained enough knowledge about valid `cborg` encodings to help us with this.

Reply:
Pros: It is implemented, and tested.

I was a bit sloppy when I described `0xa1` above.
It means: `map with 1 pair of items follow`.

Reply:
> It means: map with 1 pair of items follow.
@karknu you're not right, check my comment with the ghci snippet [above](https://github.com/input-output-hk/ouroboros-network/pull/352#discussion_r266441263) which I took from `encodeVersionMap`.

Reply:
From https://tools.ietf.org/html/rfc7049

` 0xa0..0xb7      | map (0x00..0x17 pairs of data items follow) `

Reply:
so the encoding we use here is wrong; we have `0xa0` followed by data items rather pairs of data items.

Reply:
The example I showed used `0xa1`, which was for a message with a single version. For two versions we would use `0xa2`. The map sizes are encoded by `CBOR.encodeMapLen`.
The `0xa0` is a quote from the rfc.

Example of a request with two version (from version negotiation test):
`0x82 0x00 0xa2 0x00 0x1a 0xca 0xfe 0xbe 0xeb 0x01 0x1a 0xbe 0xeb 0xca 0xfe`
0x82 Array with two Items
0x00 MsgInitReq key
0xa2 Map with 2 pairs of data items
0x00 Version 0
0x1a Unsigned Integer follows
0xcafebeeb NetworkMagic
0x01 Version 1
0x1a Unsigned Integer follows
0xbeebcafe NetworkMagic

Reply:
So the pairs you are putting there are the version plus network magic not the one in the type `Map VersionNumber Version`.  This is confusing. I would expect it all to be just:
```
encodeVersions :: [Version] -> CBOR.Encoding
encodeVersions versions =
    CBOR.encodeListLen (fromIntegral $ length versions)
    <> foldr (\v b -> encode v <> b) mempty versions

decodeCtrlMsg :: HasCallStack => CBOR.Decoder s ControlMsg
decodeCtrlMsg = do
    _ <- CBOR.decodeListLen
    key <- CBOR.decodeWord
    case key of
         0 -> MsgInitReq <$> decodeVersions
         1 -> MsgInitRsp <$> decode
         2 -> MsgInitFail <$> decode
         a -> throw $ MuxError MuxControlUnknownMessage
                      ("unknown control message type " ++ show a)
                      callStack

  where
    decodeVersions :: CBOR.Decoder s [Version]
    decodeVersions = do
        n <- CBOR.decodeListLen
        fmap catMaybes (replicateM n tryDecodeVersion)

    tryDecodeVersion :: CBOR.Decoder s (Maybe Version)
    tryDecodeVersion = do
      v <- CBOR.decodeWord
      n <- decode
      case v of
        0 -> return $ Just (Version0 n)
        1 -> return $ Just (Version1 n)
        _ -> return Nothing
```

This will also skip over unknown versions `tryDecodeVersion`.  Though it binds us to the format of two `Words` per version (version number + network number).  If you're not comfortable with this, then you could use `CBOR.termEncode` & `CBOR.termDecode` as in this PR.

Reply:
To make it more explicit I would also add a type:
```
data SomeVersion = SomeVersion !Version | UnknownVersion
```
And use it instead of `Maybe Version`, and I would add a `Serialise` instance for it.  One could also try to add `UnknownVersion` constructor to `Version` directly, but I would expect it not to work since we'll be forced to deal with `UnknownVersion` at call sites (if we have a few of them then this might be ok).

Reply:
The encoding/decoding works, it follows the CBOR behavior for `Map` and it can cope with unknown versions with an unknown format (that is not simply a version followed by a word).

Reply:
I know, the version above can also be adjusted for that (I just made a simplified version).  The problem with it is that it's not obvious from the type `Map VersionNumber Version` that the `VersionNumber` is thrown away and each `Version` is encoded as a `(key, val)` entry.